zanata-war/src/main/java/org/zanata/action/MergeTransAction.java

```diff
@@ -0,0 +1,230 @@
+import org.jboss.seam.ScopeType;
+import org.jboss.seam.annotations.Create;
+import org.jboss.seam.annotations.In;
+import org.jboss.seam.annotations.Name;
+import org.jboss.seam.annotations.Scope;
+import org.jboss.seam.faces.FacesMessages;
+import org.jboss.seam.international.StatusMessage;
+import org.zanata.async.handle.MergeTranslationsTaskHandle;
+import org.zanata.common.EntityStatus;
+import org.zanata.dao.ProjectDAO;
+import org.zanata.dao.ProjectIterationDAO;
+import org.zanata.i18n.Messages;
+import org.zanata.model.HProject;
+import org.zanata.model.HProjectIteration;
+import org.zanata.ui.CopyAction;
+
+/**
+ * Handles action from merge_trans_modal.xhtml
+ *
+ * @author Alex Eng <a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
+ */
+@Name("mergeTransAction")
+@Scope(ScopeType.PAGE)
+public class MergeTransAction implements Serializable, CopyAction {
+
+    @Getter
+    private String targetProjectSlug;
+
+    @Getter
+    @Setter
+    private String targetVersionSlug;
+
+    @Getter
+    private String sourceProjectSlug;
+
+    @Getter
+    @Setter
+    private String sourceVersionSlug;
+
+    @Getter
+    @Setter
+    private boolean keepExistingTranslation;
+
+    @In
+    private ProjectDAO projectDAO;
+
+    @In
+    private ProjectIterationDAO projectIterationDAO;
+
+    @In
+    private MergeTranslationsManager mergeTranslationsManager;
+
+    @In
+    private CopyTransManager copyTransManager;
+
+    @In
+    private CopyVersionManager copyVersionManager;
+
+    @In
+    private Messages msgs;
+
+    private HProjectIteration targetVersion;
+
+    private HProject sourceProject;
+
+    public void setTargetProjectSlug(String targetProjectSlug) {
+        this.targetProjectSlug = targetProjectSlug;
+
+        /**
+         * This should only true during first instantiation. See pages.xml
+         * for setter of mergeTransAction.targetProjectSlug
+         */
+        if(sourceProjectSlug == null) {
+            setSourceProjectSlug(targetProjectSlug);
+        }
+    }
+
+    public void setSourceProjectSlug(String sourceProjectSlug) {
+        if(!StringUtils.equals(this.sourceProjectSlug, sourceProjectSlug)) {
+            this.sourceProjectSlug = sourceProjectSlug;
+            sourceProject = null;
+            if (!getSourceProject().getProjectIterations().isEmpty()) {
+                sourceVersionSlug =
+                        getSourceProject().getProjectIterations().get(0)
+                                .getSlug();
+            }
+        }
+    }
+
+    public HProjectIteration getTargetVersion() {
+        if(targetVersion == null) {
+            targetVersion =
+                    projectIterationDAO.getBySlug(targetProjectSlug,
+                            targetVersionSlug);
+        }
+        return targetVersion;
+    }
+
+    public HProject getSourceProject() {
+        if(sourceProject == null) {
+            sourceProject = projectDAO.getBySlug(sourceProjectSlug);
```

Comment:
I think this will throw an exception if `sourceProjectSlug` is `null`. I think it would be worth looking at all usages of the 4 slugs in here and make sure they will be ok if the value is null.